src/components/tl/cuda/alltoallv/alltoallv_ce.c

```diff
@@ -0,0 +1,319 @@
+ucc_status_t ucc_tl_cuda_alltoallv_ce_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_status_t        status;
+
+    switch (task->alltoallv_ce.stage) {
+    case ALLTOALL_CE_STAGE_SYNC:
+        if (ucc_tl_cuda_get_sync(task) != UCC_OK) {
+            task->super.super.status = UCC_INPROGRESS;
+            return task->super.super.status;
+        }
+        status = ucc_tl_cuda_alltoallv_setup_start(task);
+        if (status != UCC_OK) {
+            task->super.super.status = status;
+            return task->super.super.status;
+        }
+        task->alltoallv_ce.stage = ALLTOALL_CE_STAGE_SETUP;
+    case ALLTOALL_CE_STAGE_SETUP:
+        status = ucc_tl_cuda_alltoallv_setup_test(task);
+        if (status != UCC_OK) {
+            task->super.super.status = status;
+            return task->super.super.status;
+        }
+        status = ucc_tl_cuda_alltoallv_ce_post_copies(task);
+        if (ucc_unlikely(status != UCC_OK)) {
+            task->super.super.status = status;
+            return task->super.super.status;
+        }
+        task->alltoallv_ce.stage = ALLTOALL_CE_STAGE_COPY;
+    case ALLTOALL_CE_STAGE_COPY:
+        status =
+            ucc_ec_event_test(task->alltoallv_ce.copy_done, UCC_EE_CUDA_STREAM);
+        if (status != UCC_OK) {
+            task->super.super.status = status;
+            return task->super.super.status;
+        }
+        status =
+            ucc_tl_cuda_shm_barrier_start(UCC_TL_TEAM_RANK(team), task->bar);
+        if (ucc_unlikely(status != UCC_OK)) {
+            task->super.super.status = status;
+            return task->super.super.status;
+        }
+        task->alltoallv_ce.stage = ALLTOALL_CE_STAGE_BAR;
+    default:
+        ucc_assert(task->alltoallv_ce.stage == ALLTOALL_CE_STAGE_BAR);
+        break;
+    }
+    task->super.super.status =
+        ucc_tl_cuda_shm_barrier_test(UCC_TL_TEAM_RANK(team), task->bar);
+    if (task->super.super.status == UCC_OK) {
+        ucc_tl_cuda_put_sync(task);
+    }
+    return task->super.super.status;
+}
+
+ucc_status_t ucc_tl_cuda_alltoallv_ce_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+
+    task->alltoallv_ce.stage = ALLTOALL_CE_STAGE_SYNC;
+    ucc_tl_cuda_alltoallv_ce_progress(coll_task);
+    if (task->super.super.status == UCC_INPROGRESS) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+    return ucc_task_complete(coll_task);
+}
+
+ucc_status_t ucc_tl_cuda_alltoallv_ce_init(ucc_tl_cuda_task_t *task)
+{
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_coll_args_t    *args = &TASK_ARGS(task);
+    ucc_status_t        status;
+    size_t              data_len;
+
+    if (!(args->flags & (UCC_COLL_ARGS_FLAG_CONTIG_SRC_BUFFER |
```

Comment:
i think the check is not correct:
1. you also need to check if args->mask & UCC_COLL_ARGS_FIELD_FLAGS is there (otherwise flags can have garbage)
2. you check would pass if either src or dst contig is set, while you need both

I would add helper macros to ucc_Coll_utils:
UCC_ARGS_SRC_CONTIG(args) (args.mask & UCC_COLL_ARGS_FIELD_FLAGS) && (args.flags & UCC_COLL_ARGS_FLAG_CONTIG_SRC_BUFFER)
similar for dst
and
for both:
UCC_COLL_ARGS_CONTIG_BUFFER(args) UCC_COLL_ARGS_SRC_CONTIG(args) && UCC_COLL_ARGS_DST_CONTIG(args)